Fix names of some reference functions

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88117
* #88116
* __->__ #88115

The `__name__` field of some binary reference functions was wrong. We
fix this to be consistent with unary reference functions. In the future,
we should probably make the binary reference wrapper return a wrapper
itself to avoid all those calls to `partial`.

This change helps performing some homogeneous treatment of functions by
their name.

cc @ezyang @mruberry @ngimel @Lezcano @peterbell10 @fdrocha

lezcano committed:
Fix names of some reference functions

The `__name__` field of some binary reference functions was wrong. We
fix this to be consistent with unary reference functions. In the future,
we should probably make the binary reference wrapper return a wrapper
itself to avoid all those calls to `partial`.

This change helps performing some homogeneous treatment of functions by
their name.

[ghstack-poisoned]

ezyang commented:
See also https://github.com/pytorch/pytorch/pull/79178

ezyang commented:
Well, now that you bashed out all the partials, I guess we can take it, but I probably would have added the decorator variant first haha

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -944,47 +949,70 @@ def add(
 # TODO: add docstring
-atan2 = _make_elementwise_binary_reference(
-    prims.atan2,  # type: ignore[has-type]
+@partial(
```
I don't love the pattern of `@partial` for this and then having to basically redefine the prim in a wrapper. Are there some other options?

lezcano replied:
I'll remove all the partials in this PR as well.

lezcano commented:
@ezyang vim macros FTW :D

lezcano committed:
Update on "Fix names of some reference functions"


The `__name__` field of some binary reference functions was wrong. We
fix this to be consistent with unary reference functions. In the future,
we should probably make the binary reference wrapper return a wrapper
itself to avoid all those calls to `partial`.

This change helps performing some homogeneous treatment of functions by
their name.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Fix names of some reference functions"


The `__name__` field of some binary reference functions was wrong. We
fix this to be consistent with unary reference functions. In the future,
we should probably make the binary reference wrapper return a wrapper
itself to avoid all those calls to `partial`.

This change helps performing some homogeneous treatment of functions by
their name.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Fix names of some reference functions"


The `__name__` field of some binary reference functions was wrong. We
fix this to be consistent with unary reference functions. In the future,
we should probably make the binary reference wrapper return a wrapper
itself to avoid all those calls to `partial`.

This change helps performing some homogeneous treatment of functions by
their name.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Fix names of some reference functions"


The `__name__` field of some binary reference functions was wrong. We
fix this to be consistent with unary reference functions. In the future,
we should probably make the binary reference wrapper return a wrapper
itself to avoid all those calls to `partial`.

This change helps performing some homogeneous treatment of functions by
their name.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Fix names of some reference functions"


The `__name__` field of some binary reference functions was wrong. We
fix this to be consistent with unary reference functions. In the future,
we should probably make the binary reference wrapper return a wrapper
itself to avoid all those calls to `partial`.

This change helps performing some homogeneous treatment of functions by
their name.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

mruberry commented on torch/_prims_common/__init__.py:
```diff
@@ -1576,6 +1576,27 @@ def mask_tensor(mask: TensorLikeType, t: TensorLikeType):
 
+def get_aten_op(fn: Callable, name: str):
+    """
+    Given the __module__ of reference and its name, it returns
```
Isn't there a reference->ATen mapping that this should be using?

lezcano replied:
I have no idea. I don't think such a thing exists. Perhaps @ezyang knows?

ezyang replied:
Whenever a ref is registered as a decomposition for an aten op (`register_decomposition`) that also sets up a ref to aten mapping. We don't publish this API but the info is there inside a hash table.

lezcano replied:
Yep, that's what I thought. This function is meant to be called in order to then register the relevant decomposition later on, so we can't make use of that.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -860,54 +855,59 @@ def trunc(a):
-            raise ValueError(
-                "Received a lhs Python scalar to an elementwise binary operation that does not accept lhs scalars!"
-            )
+    def inner(prim: Callable):
```
What's going on with this change?

lezcano replied:
This just aligns `_make_elementwise_binary_reference` with `_make_elementwise_unary_reference` in that, given some parameters, it returns a decorator.

lezcano replied:
I also moved from if nots to checks, improve the error messages (priting the correct name) and I set the name of the `ref`

lezcano committed:
Update on "Fix names of some reference functions"


The `__name__` field of some binary reference functions was wrong. We
fix this to be consistent with unary reference functions. In the future,
we should probably make the binary reference wrapper return a wrapper
itself to avoid all those calls to `partial`.

This change helps performing some homogeneous treatment of functions by
their name.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "Fix names of some reference functions"


The `__name__` field of some binary reference functions was wrong. We
fix this to be consistent with unary reference functions. In the future,
we should probably make the binary reference wrapper return a wrapper
itself to avoid all those calls to `partial`.

This change helps performing some homogeneous treatment of functions by
their name.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

mruberry reviewed (approved):
OK

lezcano committed:
Update on "Fix names of some reference functions"


The `__name__` field of some binary reference functions was wrong. We
fix this to be consistent with unary reference functions. In the future,
we should probably make the binary reference wrapper return a wrapper
itself to avoid all those calls to `partial`.

This change helps performing some homogeneous treatment of functions by
their name.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]
